sqlmesh/core/engine_adapter/clickhouse.py

```diff
@@ -296,6 +293,70 @@ def _build_partitioned_by_exp(
             this=exp.Schema(expressions=partitioned_by),
         )
 
+    def inject_query_setting(
+        self,
+        query: Query,
+        setting_name: t.Optional[str] = "join_use_nulls",
+        setting_value: t.Optional[str] = "1",
+        check_server_default: bool = False,
+    ) -> Query:
+        # Set a setting in the query's SETTINGS clause
+        #
+        # Use in SCD models:
+        #  - The SCD query we build must include the setting `join_use_nulls = 1` to ensure that empty cells in a join
+        #      are filled with NULL instead of the default data type value. The default join_use_nulls value is `0`.
+        #  - The SCD embeds the user's original query in the `source` CTE
+        #  - Settings are dynamically scoped, so our setting may override the server's default setting the user expects
+        #      for their query.
+        #  - To prevent this, we:
+        #     - Determine if the user query already sets `join_use_nulls` - if so, we leave it as-is
+        #     - If the user query does not set `join_use_nulls`, we query the server for the current setting. If the
+        #         current setting is `0`, we add `join_use_nulls = 0` to the user query.
+        #  - We specify our setting on the outermost scope, so it applies to all CTEs/subqueries other than the user's
+        if setting_name and setting_value:
+            inject_setting = True
+            if check_server_default:
+                user_settings = query.args.get("settings")
```

Comment:
Are we searching the whole AST in places where this method is called? I'm thinking that the `SETTINGS` clause applies to the whole query, so if it can be nested deeper than the top-level query we may miss it.

In this case, perhaps a `(select.args.get("settings") for select in find_all(Select))` generator could be used. Though there's some overhead to fetch every `Select` which is crappy...

Reply:
Hmm...a user query could contain a subquery with a different setting, but that setting would have to be explicit (in which case our setting wouldn't overwrite it).

My logic is:
- If a scope sets its own value, our value will not override
- Any scope without its own value will search "outward" to determine its value
- Therefore, any nested scopes will search until they either hit a user-specified value or the outermost user scope where we add the server default. 

For example:
```
SELECT 
  * 
FROM (
  SELECT 
    * 
  FROM inner_left 
  LEFT JOIN inner_right 
  ON inner_left.id = inner_right.id 
  SETTINGS join_use_nulls = 1
) as outer_left
LEFT JOIN outer_right
ON outer_left.id = outer_right.id
```

Assume the server value is join_use_nulls=0, so we need to inject that value to preserve this query's outer scope behavior.

Adding join_use_nulls=0 to the user's outer query won't override the 1 in the subquery, so they will see the expected behavior.